File: src/ts/src/ads/internal/DefaultBannerAd.ts
import {
    ILogger,
    IMessageBridge,
    ObserverManager,
} from "../../core";
import { AdObserver } from "../IAd";
import { IBannerAd } from "../IBannerAd";
import { AsyncHelper } from "./AsyncHelper";
import { BannerAdHelper } from "./BannerAdHelper";
import { Capper } from "./Capper";
import { IAsyncHelper } from "./IAsyncHelper";
import { ICapper } from "./ICapper";
import { MessageHelper } from "./MessageHelper";

type Destroyer = () => void;

export class DefaultBannerAd extends ObserverManager<AdObserver> implements IBannerAd {
    private readonly kTag = `DefaultBannerAd`;
    private readonly _prefix: string;
    private readonly _bridge: IMessageBridge;
    private readonly _logger: ILogger;
    private readonly _destroyer: Destroyer;
    private readonly _network: string;
    private readonly _adId: string;
    private readonly _messageHelper: MessageHelper;
    private readonly _helper: BannerAdHelper;
    private readonly _loadCapper: ICapper;
    private readonly _loader: IAsyncHelper<boolean>;

    public constructor(
        prefix: string,
        bridge: IMessageBridge,
        logger: ILogger,
        destroyer: Destroyer,
        network: string,
        adId: string,
        size: [number, number]) {
        super();
        this._prefix = prefix;
        this._bridge = bridge;
        this._logger = logger;
        this._destroyer = destroyer;
        this._network = network;
        this._adId = adId;
        this._messageHelper = new MessageHelper(prefix, adId);
        this._helper = new BannerAdHelper(this._bridge, this._messageHelper, size);
        this._loadCapper = new Capper(5);
        this._loader = new AsyncHelper<boolean>();

        this._logger.debug(`${this.kTag}: constructor: prefix = ${this._prefix} id = ${this._adId}`);
        this._bridge.registerHandler(_ => this.onLoaded(), this._messageHelper.onLoaded);
        this._bridge.registerHandler(message => {
            const json: {
                code: number,
                message: string,
            } = JSON.parse(message);
            this.onFailedToLoad(json.code, json.message);
        }, this._messageHelper.onFailedToLoad);
        this._bridge.registerHandler(_ => this.onClicked(), this._messageHelper.onClicked);
    }

    public destroy(): void {
        this._logger.debug(`${this.kTag}: destroy: prefix = ${this._prefix} id = ${this._adId}`);
        this._bridge.deregisterHandler(this._messageHelper.onLoaded);
        this._bridge.deregisterHandler(this._messageHelper.onFailedToLoad);
        this._bridge.deregisterHandler(this._messageHelper.onClicked);
        this._destroyer();
    }

    public get isLoaded(): boolean {
        return this._helper.isLoaded;
    }

    public async load(): Promise<boolean> {
        this._logger.debug(`${this.kTag}: load: prefix = ${this._prefix} id = ${this._adId} loading = ${this._loader.isProcessing}`);
        if (this._loadCapper.isCapped) {
            return false;
        }
        this._loadCapper.cap();
        return await this._loader.process(
            () => this._helper.load(),
            result => {
                // OK.
            });
    }

    public get anchor(): [number, number] {
        return this._helper.anchor;
    }

    public set anchor(value: [number, number]) {
        this._helper.anchor = value;
    }

    public get position(): [number, number] {
        return this._helper.anchor;
    }

    public set position(value: [number, number]) {
        this._helper.position = value;
    }

    public get size(): [number, number] {
        return this._helper.anchor;
    }

    public set size(value: [number, number]) {
        this._helper.size = value;
    }

    public get isVisible(): boolean {
        return this._helper.isVisible;
    }

    public set isVisible(value: boolean) {
        this._helper.isVisible = value;
    }

    private onLoaded(): void {
        this._logger.debug(`${this.kTag}: onLoaded: prefix = ${this._prefix} id = ${this._adId} loading = ${this._loader.isProcessing}`);
        if (this._loader.isProcessing) {
            this._loader.resolve(true);
            this.dispatchEvent(observer => observer.onLoadResult && observer.onLoadResult({
                network: this._network,
                result: true,
                errorCode: 0,
                errorMessage: ``,
            }));
        } else {
            // Ignored.
        }
        this.dispatchEvent(observer => observer.onLoaded && observer.onLoaded());
    }

    private onFailedToLoad(code: number, message: string): void {
        this._logger.debug(`${this.kTag}: onFailedToLoad: prefix = ${this._prefix} id = ${this._adId} loading = ${this._loader.isProcessing} message = ${message}`);
        if (this._loader.isProcessing) {
            this._loader.resolve(false);
            this.dispatchEvent(observer => observer.onLoadResult && observer.onLoadResult({
                network: this._network,
                result: false,
                errorCode: code,
                errorMessage: message,
            }));
        } else {
            // Ignored.
        }
    }

    private onClicked(): void {
        this._logger.debug(`${this.kTag}: onClicked: prefix = ${this._prefix} id = ${this._adId}`);
        this.dispatchEvent(observer => observer.onClicked && observer.onClicked());
    }
}

File: src/unity/Runtime/Ads/Internal/DefaultBannerAd.cs
using System;
using System.Threading.Tasks;

using Jsonite;

namespace EE.Internal {
    using Destroyer = Action;

    internal class DefaultBannerAd : ObserverManager<AdObserver>, IBannerAd {
        private const string kTag = nameof(DefaultBannerAd);
        private readonly string _prefix;
        private readonly IMessageBridge _bridge;
        private readonly Destroyer _destroyer;
        private readonly ILogger _logger;
        private readonly string _network;
        private readonly string _adId;
        private readonly MessageHelper _messageHelper;
        private readonly BannerAdHelper _helper;
        private readonly ICapper _loadCapper;
        private readonly IAsyncHelper<bool> _loader;

        public DefaultBannerAd(
            string prefix,
            IMessageBridge bridge,
            ILogger logger,
            Destroyer destroyer,
            string network,
            string adId,
            (int, int) size) {
            _prefix = prefix;
            _bridge = bridge;
            _logger = logger;
            _destroyer = destroyer;
            _network = network;
            _adId = adId;
            _messageHelper = new MessageHelper(prefix, adId);
            _helper = new BannerAdHelper(_bridge, _messageHelper, size);
            _loadCapper = new Capper(5);
            _loader = new AsyncHelper<bool>();

            _logger.Debug($"{kTag}: constructor: prefix = {_prefix} id = {_adId}");
            _bridge.RegisterHandler(_ => OnLoaded(), _messageHelper.OnLoaded);
            _bridge.RegisterHandler(message => {
                var json = (JsonObject) Json.Deserialize(message);
                OnFailedToLoad((int) json["code"], (string) json["message"]);
            }, _messageHelper.OnFailedToLoad);
            _bridge.RegisterHandler(_ => OnClicked(), _messageHelper.OnClicked);
        }

        public void Destroy() {
            _logger.Debug($"{kTag}: {nameof(Destroy)}: prefix = {_prefix} id = {_adId}");
            _bridge.DeregisterHandler(_messageHelper.OnLoaded);
            _bridge.DeregisterHandler(_messageHelper.OnFailedToLoad);
            _bridge.DeregisterHandler(_messageHelper.OnClicked);
            _destroyer();
        }

        public bool IsLoaded => _helper.IsLoaded;

        public async Task<bool> Load() {
            _logger.Debug($"{kTag}: {nameof(Load)}: prefix = {_prefix} id = {_adId} loading = {_loader.IsProcessing}");
            if (_loadCapper.IsCapped) {
                return false;
            }
            _loadCapper.Cap();
            return await _loader.Process(
                () => _helper.Load(),
                result => {
                    // OK.
                });
        }

        public bool IsVisible {
            get => _helper.IsVisible;
            set => _helper.IsVisible = value;
        }

        public (float, float) Anchor {
            get => _helper.Anchor;
            set => _helper.Anchor = value;
        }

        public (float, float) Position {
            get => _helper.Position;
            set => _helper.Position = value;
        }

        public (float, float) Size {
            get => _helper.Size;
            set => _helper.Size = value;
        }

        private void OnLoaded() {
            _logger.Debug(
                $"{kTag}: {nameof(OnLoaded)}: prefix = {_prefix} id = {_adId} loading = {_loader.IsProcessing}");
            if (_loader.IsProcessing) {
                _loader.Resolve(true);
                DispatchEvent(observer => observer.OnLoadResult?.Invoke(new AdLoadResult {
                    Network = _network,
                    Result = true
                }));
            } else {
                // Ignored.
            }
            DispatchEvent(observer => observer.OnLoaded?.Invoke());
        }

        private void OnFailedToLoad(int code, string message) {
            _logger.Debug(
                $"{kTag}: {nameof(OnFailedToLoad)}: prefix = {_prefix} id = {_adId} loading = {_loader.IsProcessing} message = {message}");
            if (_loader.IsProcessing) {
                _loader.Resolve(false);
                DispatchEvent(observer => observer.OnLoadResult?.Invoke(new AdLoadResult {
                    Network = _network,
                    Result = true,
                    ErrorCode = code,
                    ErrorMessage = message
                }));
            } else {
                // Ignored.
            }
        }

        private void OnClicked() {
            _logger.Debug($"{kTag}: {nameof(OnClicked)}: prefix = {_prefix} id = {_adId}");
            DispatchEvent(observer => observer.OnClicked?.Invoke());
        }
    }
}